hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/SCMPipelineManager.java

```diff
@@ -379,6 +397,44 @@ public void deactivatePipeline(PipelineID pipelineID)
     stateManager.deactivatePipeline(pipelineID);
   }
 
+  /**
+   * Wait a pipeline to be OPEN.
+   *
+   * @param pipelineID ID of the pipeline to wait for.
+   * @param timeout    wait timeout, millisecond, 0 to use default value
+   * @throws IOException in case of any Exception, such as timeout
+   */
+  @Override
+  public void waitPipelineReady(PipelineID pipelineID, long timeout)
+      throws IOException {
+    Pipeline pipeline;
+    try {
+      pipeline = stateManager.getPipeline(pipelineID);
+    } catch (PipelineNotFoundException e) {
+      throw new PipelineNotFoundException(String.format(
+          "Pipeline %s cannot be found", pipelineID));
+    }
+
+    boolean ready;
+    long st = Time.monotonicNow();
+    if (timeout == 0) {
+      timeout = pipelineWaitDefaultTimeout;
+    }
+    for(ready = pipeline.isOpen();
```

Comment:
As the Pipeline object is immutable we will need to get the pipeline from stateManager every time.

Reply:
Good point.  New patch is ready.